phoenix-core-client/src/main/java/org/apache/phoenix/schema/ConditionalTTLExpression.java

```diff
@@ -0,0 +1,501 @@
+import org.apache.phoenix.schema.tuple.MultiKeyValueTuple;
+import org.apache.phoenix.schema.types.PBoolean;
+import org.apache.phoenix.thirdparty.com.google.common.collect.Lists;
+import org.apache.phoenix.thirdparty.com.google.common.collect.Sets;
+import org.apache.phoenix.util.CDCUtil;
+import org.apache.phoenix.util.EnvironmentEdgeManager;
+import org.apache.phoenix.util.SchemaUtil;
+import org.apache.phoenix.util.ViewUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ConditionalTTLExpression extends TTLExpression {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ConditionalTTLExpression.class);
+
+    // expression as passed in the DDL statement and stored in syscat
+    private final String ttlExpr;
+    // compiled expression according to the table schema. For indexes the expression is
+    // first re-written to use index column references and then compiled.
+    private final Expression compiledExpr;
+    // columns referenced in the ttl expression to be added to scan
+    private final Set<ColumnReference> conditionExprColumns;
+
+    public ConditionalTTLExpression(String ttlExpr) {
+        this.ttlExpr = ttlExpr;
+        this.compiledExpr = null;
+        this.conditionExprColumns = null;
+    }
+
+    private ConditionalTTLExpression(String ttlExpr,
+                                     Expression compiledExpression,
+                                     Set<ColumnReference> conditionExprColumns) {
+        this.ttlExpr = ttlExpr;
+        this.compiledExpr = compiledExpression;
+        this.conditionExprColumns = conditionExprColumns;
+    }
+
+    public ConditionalTTLExpression(ConditionalTTLExpression expr) {
+        this.ttlExpr = expr.ttlExpr;
+        this.compiledExpr = null;
+        this.conditionExprColumns = null;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        ConditionalTTLExpression that = (ConditionalTTLExpression) o;
+        return ttlExpr.equals(that.ttlExpr);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(ttlExpr);
+    }
+
+    @Override
+    public String getTTLExpression() {
+        return ttlExpr;
+    }
+
+    @Override
+    public String toString() {
+        return getTTLExpression();
+    }
+
+    /**
+     * The cells of the row (i.e., result) read from HBase store are lexicographically ordered
+     * for tables using the key part of the cells which includes row, family, qualifier,
+     * timestamp and type. The cells belong of a column are ordered from the latest to
+     * the oldest. The method leverages this ordering and groups the cells into their columns
+     * based on the pair of family name and column qualifier.
+     */
+    private List<Cell> getLatestRowVersion(List<Cell> result) {
```

Comment:
The TTL region scanner should not call this if the scan is not a raw scan as this computation is not necessary in that case.